iothub/device/src/InternalClient.cs

```diff
@@ -1071,37 +1072,39 @@ public async Task SetDesiredPropertyUpdateCallbackAsync(DesiredPropertyUpdateCal
 
         /// <summary>
         /// Set a callback that will be called whenever the client receives a state update
-        /// (desired or reported) from the service.  This has the side-effect of subscribing
-        /// to the PATCH topic on the service.
+        /// (desired or reported) from the service.
+        /// Set callback value to null to clear.
         /// </summary>
+        /// <remarks>
+        /// This has the side-effect of subscribing to the PATCH topic on the service.
+        /// </remarks>
         /// <param name="callback">Callback to call after the state update has been received and applied</param>
         /// <param name="userContext">Context object that will be passed into callback</param>
         /// <param name="cancellationToken">A cancellation token that can be used by other objects or threads to receive notice of cancellation.</param>
         public async Task SetDesiredPropertyUpdateCallbackAsync(DesiredPropertyUpdateCallback callback, object userContext, CancellationToken cancellationToken)
         {
-            if (callback == null)
-            {
-                throw new ArgumentNullException(nameof(callback));
-            }
-
             // Codes_SRS_DEVICECLIENT_18_003: `SetDesiredPropertyUpdateCallbackAsync` shall call the transport to register for PATCHes on it's first call.
             // Codes_SRS_DEVICECLIENT_18_004: `SetDesiredPropertyUpdateCallbackAsync` shall not call the transport to register for PATCHes on subsequent calls
-            if (!_twinPatchSubscribedWithService)
+            try
             {
-                try
+                if (callback != null && !_twinPatchSubscribedWithService)
                 {
                     await InnerHandler.EnableTwinPatchAsync(cancellationToken).ConfigureAwait(false);
                     _twinPatchSubscribedWithService = true;
                 }
-                catch (IotHubCommunicationException ex) when (ex.InnerException is OperationCanceledException)
+                else if (callback == null && _twinPatchSubscribedWithService)
                 {
-                    cancellationToken.ThrowIfCancellationRequested();
-                    throw;
+                    await InnerHandler.DisableTwinPatchAsync(cancellationToken).ConfigureAwait(false);
```

Comment:
It looks like `EnableTwinPatchAsync` and `DisableTwinPatchAsync` both delegate the operation to the pipeline/ transport layer directly. Is that where you check if enabling/ disabling is required (enable only for the first time, disable only if it has been previously enabled)?

Reply:
correct. we have a bool for this check.

Reply:
got it!